File: cpp/GlobalPosition.js
/*! \file globalposition.cpp
 *  The GlobalPosition class, along with the DateTime class,  provide a C++
 *  wrapper for the Calender-Date-Time Library in cdtlib.c.
 */

export class GlobalPosition {
  /**
   * Constructs a new GlobalPosition instance with the passed values.
   *
   * @param {number} latitude Latitude in degrees east (+) or west (-) of Greenwich
   * @param {number} longitude Longitude in degrees north (+) or south (-) of equator
   * @param {number} gmtDiff Local time difference from GMT in hours (localTime = gmtDiff + GMT)
   * @param {string} locationName Arbitrary location name
   * @param {zoneName} zoneName Time zone name
   * @returns {GlobalPosition} Reference to *this*.
   */
  constructor (lat, lon, gmtDiff = 0, zoneName = '', locationName = '') {
    this._lat = lat
    this._lon = lon
    this._gmt = gmtDiff
    this._locationName = locationName
    this._zoneName = zoneName
  }

  /**
   * Returns the position's time difference (hours) from GMT
   * @returns {number} Local time difference from GMT in hours.
   */
  gmtDiff () { return this._gmt }

  /**
   * Sets the position's local time difference from GMT.
   *
   * @param {number} hours Local time difference from GMT in hours.
   * @returns {GlobalPosition} Reference to *this*.
   */
  setGmtDiff (hours) {
    this._gmt = hours
    return this
  }

  /**
   * Gets the position's latitude.
   *
   * @returns {number} Latitude in degrees north (+) or south (-) of the equator.
   */
  latitude () { return this._lat }

  /**
   * Sets the position's latitude.
   *
   * @param {number} latitude Latitude in degrees north (+) or south (-) of the equator.
   * @returns {GlobalPosition} Reference to *this*.
   */
  setLatitude (degrees) {
    this._lat = degrees
    return this
  }

  /**
   * Gets the position's location name.
   * @returns {string} The position location name.
   */
  locationName () { return this._locationName }

  /**
   * Sets the position's location name.
   *
   * @param {string} locationName New location name
   * @returns {GlobalPosition} Reference to *this*.
   */
  setLocationName (locationName) {
    this._locationName = locationName
    return this
  }

  /**
   * Gets the position's longtude.
   *
   * @returns {number} Longitude in degrees east (+) or west (-) of Greenwich Meridian.
   */
  longitude () { return this._lon }

  /**
   * Sets the position's longitude.
   *
   * @param {number} longitude Longitude in degrees east (+) or west (-) of Greenwich Meridian.
   * @returns {GlobalPosition} Reference to *this*.
   */
  setLongitude (degrees) {
    this._lon = degrees
    return this
  }

  /**
 * Sets all the position's properties.
 *
 * @param {number} latitude Latitude in degrees east (+) or west (-) of Greenwich
 * @param {number} longitude Longitude in degrees north (+) or south (-) of equator
 * @param {number} gmtDiff Local time difference from GMT in hours (localTime = gmtDiff + GMT)
 * @param {string} locationName Arbitrary location name
 * @param {zoneName} zoneName Time zone name
   * @returns {GlobalPosition} Reference to *this*.
 */
  setPosition (longitude, latitude, gmtDiff, zoneName = null, locationName = null) {
    this._lon = longitude
    this._lat = latitude
    this._gmt = gmtDiff
    this._zoneName = zoneName === null ? this._zoneName : zoneName
    this._locationName = locationName === null ? this._locationName : locationName
    return this
  }

  /**
   * Gets the position's time zone name.
   * @returns {string} The time zone name
   */
  timeZoneName () { return this._zoneName }

  /**
   * Sets the position's time zone name
   *
   * @param {string} zoneName New time zone name
   * @returns {GlobalPosition} Reference to *this*.
   */
  setTimeZoneName (zoneName) {
    this._zoneName = zoneName
    return this
  }

  /**
 * Convenience routine to convert integral degrees, minutes, and seconds into decimal degrees.
 * No sign adjustment is made for east/west or north/south.
 *
 * @param {integer} degrees arc degrees
 * @param {integer} minutes degree arc-minutes
 * @param {number} seconds degree arc seconds
 * @returns {number} Decimal degrees.
 */
  dmsToDeg (degrees, minutes, seconds) { return degrees + minutes / 60 + seconds / 3600 }

  /**
 * Convenience routine to convert decimal degrees to integral degrees, minutes, and seconds.
 * No sign conversion is made for north/south or east/west.
 *
 * @param {number} Decimal degrees
 * @returns {array} [integerDegrees, integerMinutes, floatSeconds]
 */
  degToDms (decimal) {
    let d = Math.abs(decimal)
    const degrees = Math.floor(d)
    d = 60 * (d - degrees)
    const minutes = Math.floor(d)
    const seconds = 60 * (d - minutes)
    return [degrees, minutes, seconds]
  }
}

File: cpp/globalPosition.cpp
//------------------------------------------------------------------------------
/*! \file globalposition.cpp
 *  \version BehavePlus3
 *  \author Copyright (C) 2002-2018 by Collin D. Bevins.  All rights reserved.
 *
 *  \brief Global position C++ source code.
 *
 *  The GlobalPosition class, along with the DateTime class,  provide a C++
 *  wrapper for the Calender-Date-Time Library in cdtlib.c.
 */

// Custom include files
#include "globalposition.h"

// Qt include files
#include <qstring.h>

// Standard include files
#include <math.h>
#include <stdio.h>

//------------------------------------------------------------------------------
/*! \brief Constructs a new GlobalPosition instance with default values.
 *
 *  \arg m_lat = 0 degrees (equator).
 *  \arg m_lon = 0 degrees (Greenwich Meridian).
 *  \arg m_gmt = 0 hours
 */

GlobalPosition::GlobalPosition( void ) :
    m_locationName(""),
    m_zoneName(""),
    m_lat(0.),
    m_lon(0.),
    m_gmt(0.)
{
    return;
}

//------------------------------------------------------------------------------
/*! \brief GlobalPosition copy constructor.
 *
 *  \param gp Reference to an existing GlobalPosition.
 */

GlobalPosition::GlobalPosition( GlobalPosition &gp ) :
    m_locationName(gp.m_locationName),
    m_zoneName(gp.m_zoneName),
    m_lat(gp.m_lat),
    m_lon(gp.m_lon),
    m_gmt(gp.m_gmt)
{
    return;
}

//------------------------------------------------------------------------------
/*! \brief Constructs a new GlobalPosition instance with the passed values.
 *
 *  Latitudes \e north of the equator have \e positive values, and
 *  latitudes \e south of the equator have \e negative values.
 *
 *  Longitudes \e west of Greenwich Meridian have \e positive values, and
 *  longitudes \e east of the Greenwich Meridian have \e negative values.
 *
 *  The \a gmtDiff is the number of hours added to GMT to obtain the local
 *  time ( LocalTime = gmtDiff + GMT ).
 *
 *  \param longitude Longitude in degrees.
 *  \param latitude Latitude in degrees.
 *  \param gmtDiff Local time difference from GMT in hours.
 */

GlobalPosition::GlobalPosition( double longitude, double latitude,
        double gmtDiff ) :
    m_locationName(""),
    m_zoneName(""),
    m_lat(latitude),
    m_lon(longitude),
    m_gmt(gmtDiff)
{
    return;
}

//------------------------------------------------------------------------------
/*! \brief Constructs a new GlobalPosition instance with the passed values.
 *
 *  Latitudes \e north of the equator have \e positive values, and
 *  latitudes \e south of the equator have \e negative values.
 *
 *  Longitudes \e west of Greenwich Meridian have \e positive values, and
 *  longitudes \e east of the Greenwich Meridian have \e negative values.
 *
 *  The \a gmtDiff is the number of hours added to GMT to obtain the local
 *  time ( LocalTime = gmtDiff + GMT ).
 *
 *  \param locationName Geographic place name.
 *  \param zoneName Time zone name.
 *  \param longitude Longitude in degrees.
 *  \param latitude Latitude in degrees.
 *  \param gmtDiff Local time difference from GMT in hours.
 */

GlobalPosition::GlobalPosition(
        const QString &locationName, const QString &zoneName,
        double longitude, double latitude, double gmtDiff ) :
    m_locationName(locationName),
    m_zoneName(zoneName),
    m_lat(latitude),
    m_lon(longitude),
    m_gmt(gmtDiff)
{
    return;
}

//------------------------------------------------------------------------------
/*! \brief GlobalPosition destructor.
 */

GlobalPosition::~GlobalPosition( void )
{
    return;
}

//------------------------------------------------------------------------------
/*! \brief Gets the local time difference from GMT.
 *
 *  \return Local time difference from GMT in hours.
 */

double GlobalPosition::gmtDiff( void ) const
{
    return( m_gmt );
}

//------------------------------------------------------------------------------
/*! \brief Sets the local time difference from GMT.
 *
 *  \param Local time difference from GMT in hours.
 *
 *  \return New local time difference from GMT in hours.
 */

double GlobalPosition::gmtDiff( double hours )
{
    return( m_gmt = hours );
}

//------------------------------------------------------------------------------
/*! \brief Gets the position latitude.
 *
 *  Latitudes \e north of the equator have \e positive values, and
 *  latitudes \e south of the equator have \e negative values.
 *
 *  \return Position latitude in decimal degrees.
 */

double GlobalPosition::latitude( void ) const
{
    return( m_lat );
}

//------------------------------------------------------------------------------
/*! \brief Sets the position latitude.
 *
 *  Latitudes \e north of the equator have \e positive values, and
 *  latitudes \e south of the equator have \e negative values.
 *
 *  \param degrees New latitude in decimal degrees.
 *
 *  \return New position latitude in decimal degrees.
 */

double GlobalPosition::latitude( double degrees )
{
    return( m_lat = degrees );
}

//------------------------------------------------------------------------------
/*! \brief Gets the current location name.
 *
 *  \return Reference to the current location name.
 */

const QString &GlobalPosition::locationName( void ) const
{
    return( m_locationName );
}

//------------------------------------------------------------------------------
/*! \brief Sets the current location name.
 *
 *  \return Reference to the new location name.
 */

QString &GlobalPosition::locationName( const QString &name )
{
    return( m_locationName = name );
}

//------------------------------------------------------------------------------
/*! \brief Gets the position longitude.
 *
 *  Longitudes \e west of Greenwich Meridian have \e positive values, and
 *  longitudes \e east of the Greenwich Meridian have \e negative values.
 *
 *  \return Position longitude in decimal degrees.
 */

double GlobalPosition::longitude( void ) const
{
    return( m_lon );
}

//------------------------------------------------------------------------------
/*! \brief Sets the position longtiude.
 *
 *  Longitudes \e west of Greenwich Meridian have \e positive values, and
 *  longitudes \e east of the Greenwich Meridian have \e negative values.
 *
 *  \param degrees The new longitude in decimal degrees.
 *
 *  \return New position longitude in decimal degrees.
 */

double GlobalPosition::longitude( double degrees )
{
    return( m_lon = degrees );
}

//------------------------------------------------------------------------------
/*! \brief Prints the GlobalPosition member data to the FILE stream.
 *
 *  \param fptr Pointer to an open FILE stream.
 */

void GlobalPosition::print( FILE *fptr ) const
{
    fprintf( fptr,
        "Global position is %s%3.2f, %s%3.2f (GMT + %1.2f)\n",
        ( m_lon >= 0. )
            ? "W"
            : "E",
        fabs( m_lon ),
        ( m_lat >= 0. )
            ? "N"
            : "S",
        fabs( m_lat ),
        m_gmt );
    return;
}

//------------------------------------------------------------------------------
/*! \brief Sets all the protected data members.
 *
 *  Latitudes \e north of the equator have \e positive values, and
 *  latitudes \e south of the equator have \e negative values.
 *
 *  Longitudes \e west of Greenwich Meridian have \e positive values, and
 *  longitudes \e east of the Greenwich Meridian have \e negative values.
 *
 *  The \a gmtDiff is the number of hours added to GMT to obtain the local
 *  time ( LocalTime = gmtDiff + GMT ).
 *
 *  \param longitude The new longitude in decimal degrees.
 *  \param latitude The new latitude in decimal degrees.
 *  \param gmtDiff The new local time difference from GMT in  hours.
 */

void GlobalPosition::setPosition( double longitude, double latitude,
    double gmtDiff )
{
    m_lon = longitude;
    m_lat = latitude;
    m_gmt = gmtDiff;
    return;
}

//------------------------------------------------------------------------------
/*! \brief Gets the current time zone name.
 *
 *  \return Reference to the current time zone name.
 */

const QString &GlobalPosition::zoneName( void ) const
{
    return( m_zoneName );
}

//------------------------------------------------------------------------------
/*! \brief Sets the current time zone name.
 *
 *  \return Reference to the new zone name.
 */

QString &GlobalPosition::zoneName( const QString &name )
{
    return( m_zoneName = name );
}

//------------------------------------------------------------------------------
/*! \brief GlobalPosition::operator=() assignment opertor.
 *
 *  \brief Assignment operator.
 *
 *  \param gp Reference to an existing GlobalPosition.
 *
 *  \return A reference to \a this object.
 */

GlobalPosition &GlobalPosition::operator=( const GlobalPosition &gp )
{
    m_lon = gp.m_lon;
    m_lat = gp.m_lat;
    m_gmt = gp.m_gmt;
    return( *this );
}

//------------------------------------------------------------------------------
/*! \brief Convenience routine to convert integral degrees, minutes, and
 *  seconds into decimal degrees.  No sign adjustment is made for east/west or
 *  north/south.
 *
 *  \return Decimal degrees.
 */

double DmsToDeg( int degrees, int minutes, int seconds )
{
    return( (double) degrees + (double) minutes/60. + (double) seconds/3600. );
}

//------------------------------------------------------------------------------
/*! \brief Convenience routine to convert decimal degrees to integral degrees,
 *  minutes, and seconds.  No sign conversion is made for north/south or
 *  east/west.
 *
 *  \return Decimal degrees.
 */

void DegToDms( double decimal, int *degrees, int *minutes, int *seconds )
{
    decimal  = fabs( decimal );
    *degrees = (int) decimal;
    decimal  = 60. * ( decimal - *degrees );
    *minutes = (int) decimal;
    decimal  = 60. * ( decimal - *minutes );
    *seconds = (int) ( decimal + 0.5 );     // Round to nearest second
    return;
}

//------------------------------------------------------------------------------
//  End of globalposition.cpp
//------------------------------------------------------------------------------

